src/boot/OfflineNoticeProvider.js

```diff
@@ -0,0 +1,454 @@
+        <View onLayout={handleNoticeContentAreaLayoutChange} style={styles.noticeContentArea}>
+          <ZulipTextIntl
+            style={styles.noticeText}
+            text={
+              shouldShowUncertaintyNotice
+                ? 'Please check your Internet connection'
+                : 'No Internet connection'
+            }
+          />
+        </View>
+      </SafeAreaView>
+      <View style={styles.flex1}>
+        <OfflineNoticeContext.Provider value={contextValue}>
+          {props.children}
+        </OfflineNoticeContext.Provider>
+      </View>
+    </View>
+  );
+}
+
+type PlaceholderProps = $ReadOnly<{|
+  ...ViewProps,
+  style?: ViewStylePropWithout<{|
+    // Let this component do its job; don't mess with its height.
+    height: DimensionValue,
+    minHeight: DimensionValue,
+    maxHeight: DimensionValue,
+  |}>,
+|}>;
+
+/**
+ * Empty View that expands from 0 height to give room to the offline notice.
+ *
+ * On every screen, render one of these on the surface that occupies the top
+ * of the screen in the Y direction (e.g., an app bar), inside that
+ * surface's top-inset padding.
+ *
+ * The offline notice will be overlaid on top of that surface in the Z
+ * direction, blocking part of it from view including its top-inset padding.
+ * Use this placeholder to push the content on the underlying surface
+ * downward in the Y direction, just enough so it doesn't get hidden by the
+ * offline notice's content and only when the offline notice is actually
+ * onscreen.
+ *
+ * On Android, where the app doesn't draw underneath the status bar, this
+ * also colors the status bar's background to match the notice.
+ *
+ * Must have the OfflineNoticeProvider above it in the tree.
+ */
+export function OfflineNoticePlaceholder(props: PlaceholderProps): Node {
+  const { style: callerStyle } = props;
+
+  const { isNoticeVisible, noticeContentAreaHeight } = useContext(OfflineNoticeContext);
+  const prevIsNoticeVisible = usePrevious(isNoticeVisible, isNoticeVisible);
+
+  const plainHeight = isNoticeVisible ? noticeContentAreaHeight : 0;
+  const animHeight = useRef(new Animated.Value(plainHeight)).current;
+
+  // Part of an Android workaround; see where we set the View's height.
+  useConditionalEffect(
+    useCallback(() => {
+      // Should approximate OfflineNoticeProvider's animation curve.
+      const animation = Animated.timing(animHeight, {
+        toValue: plainHeight,
+        duration: isNoticeVisible ? 1000 : 300,
+        easing: Easing.inOut(t => Easing.ease(t)),
+
+        // With `true`, I get an error:
+        // - On Android: "Animated node with tag […] does not exist"
+        // - On iOS: "Style property 'height' is not supported by native
+        //   animated module".
+        useNativeDriver: false,
+      });
+      animation.start();
+      return animation.stop;
+    }, [isNoticeVisible, animHeight, plainHeight]),
+    Platform.OS === 'android' && prevIsNoticeVisible !== isNoticeVisible,
```

Comment:
Oh OK, yeah, I think `useConditionalEffect` is the wrong thing to use here.

> If you take it out (so the condition is just `Platform.OS === 'android'`), does anything go wrong?

Yes, I think so: the effect (whose job is to schedule entrance/exit animations) will run on Android when the value of the callback changes, and the value of the callback will change when `plainHeight` changes because it's in the deps we give to `useCallback`. But we don't want a change in `plainHeight` to trigger the effect; we only want to respond to changes in `isNoticeVisible`, not, e.g., system font-size changes or the initialization of `plainHeight` from the bogus value 0 to a real value.

Reply:
The new logic LGTM!